nixos/modules/services/matrix/synapse.nix

```diff
@@ -992,6 +1025,17 @@ in {
           by synapse in `services.matrix-synapse.settings.listeners` or in one of the workers!
         '';
       }
+      {
+        assertion = lib.all
+          (
+            listener:
+              listener.path != null || (listener.bind_addresses != null && listener.port != null)
+          ) cfg.settings.listeners;
+        message = ''
+          Listener configurations require either `bind_addresses` and `port` for TCP sockets,
+          or `path` for UNIX domain sockets.
+        '';
```

Comment:
It's perhaps a little easier to generate one assertion for each listener and then dump the `path` & `bind_addresses` into the message. That probably makes it easier to spot the bogus listener.

Reply:
Added an assertion, and mentioned the path. Since `bind_addresses`, `port` and `tls` are unsupported when `path` is set, and any one of them could be null, I decided not to print them.